certs/certs.go

```diff
@@ -0,0 +1,293 @@
+// instance after the last valid finality certificates, any newly finalized tipsets, and the next
+// power table to use as-of the last valid finality certificate. If passed a non-nil `base` tipset,
+// validate that the finalized chain starts with that tipset (accept any finalized chain otherwise).
+//
+// Returns an error if it encounters any invalid finality certificates, along with the last valid
+// instance, finalized chain epochs, etc. E.g., if provided a partially valid chain of finality
+// certificates, this function will return a (possibly empty) prefix of the EC chain correctly
+// finalized, the instance of the first invalid finality certificate, and the power table that
+// should be used to validate that finality certificate, along with the error encountered.
+func ValidateFinalityCertificates(verifier gpbft.Verifier, network gpbft.NetworkName, prevPowerTable gpbft.PowerEntries, nextInstance uint64, base *gpbft.TipSet,
+	certs ...FinalityCertificate) (_nextInstance uint64, chain gpbft.ECChain, newPowerTable gpbft.PowerEntries, err error) {
+	for _, cert := range certs {
+		if cert.GPBFTInstance != nextInstance {
+			return nextInstance, chain, prevPowerTable, xerrors.Errorf("expected instance %d, found instance %d", nextInstance, cert.GPBFTInstance)
+		}
+		// Some basic sanity checks.
+		if err := cert.ECChain.Validate(); err != nil {
+			return nextInstance, chain, prevPowerTable, xerrors.Errorf("invalid finality certificate at instance %d: %w", cert.GPBFTInstance, err)
+		}
+
+		// We can't have a finality certificate for "bottom"
+		if cert.ECChain.IsZero() {
+			return nextInstance, chain, prevPowerTable, xerrors.Errorf("empty finality certificate for instance %d", cert.GPBFTInstance)
+		}
+
+		// Validate that the base is as expected if specified. Otherwise, skip this check
+		// for the first finality certificate.
+		if base != nil && !base.Equal(cert.ECChain.Base()) {
+			return nextInstance, chain, prevPowerTable, xerrors.Errorf("base tipset does not match finalized chain at instance %d", cert.GPBFTInstance)
+		}
+
+		// Validate signature.
+		if err := verifyFinalityCertificateSignature(verifier, prevPowerTable, network, cert); err != nil {
+			return nextInstance, chain, prevPowerTable, err
+		}
+
+		// Now compute the new power table and validate that it matches the power table for
+		// new head.
+		newPowerTable, err = ApplyPowerTableDiff(prevPowerTable, cert.PowerTableDelta)
+		if err != nil {
+			return nextInstance, chain, prevPowerTable, xerrors.Errorf("failed to apply power table delta for finality certificate for instance %d: %w", cert.GPBFTInstance, err)
+		}
+
+		powerTableCid, err := MakePowerTableCID(newPowerTable)
+		if err != nil {
+			return nextInstance, chain, prevPowerTable, xerrors.Errorf("failed to make power table CID for finality certificate for instance %d: %w", cert.GPBFTInstance, err)
+		}
+
+		if !bytes.Equal(cert.SupplementalData.PowerTable, powerTableCid) {
```

Comment:
Nit: seeing this in use, perhaps we can make this name more obvious. E.g. SupplementalData.NextBasePowerTable? It's still tricky though...

Reply:
It's not really the next _base_ power table, it's the power table used to validate the next finality certificate. I can call it "next certificate power table" or "next instance power table"?